File: python/binary-tree-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        s = [(root, 0)]
        while s:
            node, level = s.pop()
            if node == None:
                continue
            if level == len(levels):
                levels.append([])
            levels[level].append(node.val)
            s.append((node.right, level + 1))
            s.append((node.left, level + 1))
        levels.reverse()
        return levels

File: go/binary-tree-level-order-traversal-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type PairTreeNodeLevel struct {
	Node  *TreeNode
	Level int
}

func reverseList(list [][]int) [][]int {
	length := len(list)
	for i := 0; i < (length / 2); i++ {
		list[i], list[length-i-1] = list[length-i-1], list[i]
	}
	return list
}

func levelOrderBottom(root *TreeNode) [][]int {
	levels := make([][]int, 0)
	s := make([]PairTreeNodeLevel, 0)
	s = append(s, PairTreeNodeLevel{Node: root, Level: 0})
	for 0 < len(s) {
		pair := s[0]
		s = s[1:]
		if pair.Node == nil {
			continue
		}
		if pair.Level == len(levels) {
			levels = append(levels, make([]int, 0))
		}
		levels[pair.Level] = append(levels[pair.Level], pair.Node.Val)
		s = append(s, PairTreeNodeLevel{Node: pair.Node.Left, Level: pair.Level + 1})
		s = append(s, PairTreeNodeLevel{Node: pair.Node.Right, Level: pair.Level + 1})
	}
	levels = reverseList(levels)
	return levels
}
